src/System.Management.Automation/engine/runtime/ScriptBlockToPowerShell.cs

```diff
@@ -313,6 +316,129 @@ private ScriptBlockToPowerShellConverter()
+                    {
+                        var value = Compiler.GetExpressionValue(usingAst.SubExpression, isTrustedInput, context);
+                        string usingAstKey = PsUtils.GetUsingExpressionKey(usingAst);
+                        usingValueMap.TryAdd(usingAstKey, value);
+                    }
+                }
+            }
+            catch (RuntimeException rte)
+            {
+                if (rte.ErrorRecord.FullyQualifiedErrorId.Equals("VariableIsUndefined", StringComparison.Ordinal))
+                {
+                    throw InterpreterError.NewInterpreterException(
+                        targetObject: null, 
+                        exceptionType: typeof(RuntimeException),
+                        errorPosition: usingAst.Extent, 
+                        resourceIdAndErrorId: "UsingVariableIsUndefined",
+                        resourceString: AutomationExceptions.UsingVariableIsUndefined, 
+                        args: rte.ErrorRecord.TargetObject);
+                }
+            }
+            finally
+            {
+                if (context != null)
+                {
+                    context.EngineSessionState.CurrentScope.StrictModeVersion = oldStrictVersion;
+                }
+            }
+
+            return usingValueMap;
+        }
+
+        /// <summary>
+        /// Walks the using Ast to verify it is used within a foreach-object -parallel command
+        /// and parameter set scope, and not from within a nested foreach-object -parallel call.
+        ///     $Test1 = "Hello"
+        ///     1 | ForEach-Object -Parallel { 
+        ///         $using:Test1
+        ///         $Test2 = "Goodbye"
+        ///         1 | ForEach-Object -Parallel {
+        ///             $using:Test1    # Invalid using scope
+        ///             $using:Test2    # Valid using scope
+        ///         }
+        ///     }
+        /// </summary>
+        /// <param name="usingAst">Using Ast to check.</param>
+        /// <returns>True if using expression is in current call scope.</returns>
+        private static bool IsInForeachParallelCallingScope(UsingExpressionAst usingAst)
+        {
+            Diagnostics.Assert(usingAst != null, "usingAst argument cannot be null.");
+
+            // Search up the parent Ast chain for 'Foreach-Object -Parallel' commands.
+            Ast currentParent = usingAst.Parent;
+            int foreachNestedCount = 0;
+            while (currentParent != null)
+            {
+                // Look for Foreach-Object outer commands
+                if (currentParent is CommandAst commandAst)
+                {
+                    foreach (var commandElement in commandAst.CommandElements)
+                    {
+                        if (commandElement is StringConstantExpressionAst commandName)
+                        {
+                            if (commandName.Value.Equals("foreach", StringComparison.OrdinalIgnoreCase) ||
```

Comment:
This won't work correctly for custom aliases to `ForEach-Object` or if user redefines one of these aliases to something else.

Reply:
Yeah, I didn't want to do a full command discovery look up here, as it would really hurt performance.  But maybe there is a way to do a reverse look up before walking the Ast, to find all aliases of 'foreach-object'.  I'll look into that and see if that is feasible.

Reply:
After looking at this I don't see a good way to get all possible aliases that resolve to foreach-object, without severely affecting performance.  One solution is to update sessionstate alias (scope) tables to maintain a reverse list that can be used to quickly find a list of aliases.  But this is outside the scope of this work.

For now I'll add a comment to this code, which can be revisited if we decide to add the alias reverse list table.